### ISSUE TITLE
Geuest Account

### ISSUE THREAD

**Issue by jayamin-10** [derricksmith/phpsaml#135]
@DonutsNL Hi i have downloaded the plugins from your repository. it works for the account on my domain but i can't login using a guest account on my domain. i get blocked on the this empty page. do you know if this is really achievable i mean login as a guest account ?
![avec SAML_invité](https://github.com/derricksmith/phpsaml/assets/62603670/e1df6fd6-e11a-4e98-a20e-a62cc7474cb8)

thanks in advance for your reply!

**Comment by DonutsNL** [comment]
Hi Jayamin,

Thank you for using our plugin.

**First some background :**
As far as the PHPSAML is concerned the plugin 'does not care' as long as the SAML information returned by the Identity provider (IdP) for example Azure AD is valid. In other words, the actual validation and decision to allow any account (guest or not) to login is made by the identity provider (IdP) not by our plugin. If the user is allowed the IdP should return the user browser with valid SAML body including at the very least name and email claims back tot PHPSAML that is than being processed and should login the user (if that JIT is enabled assuming these accounts are not prepopulated in GLPI). 

**The white screen**
The white screen usually is caused by an error that we did not anticipate. Could you please review the GLPI errorlogs and share any relevant (errors reported by acs)  in this feed for me to review?

Thanks!

**Comment by jayamin-10** [comment]
Hey @DonutsNL , thanks a lot for your reply. I tried to find errors reported by acs as you have said but i couldn't found any other error apart from undefined array key.
I am going to test and i will report back at you if i can login with a guest account.

**Comment by DonutsNL** [comment]
Proxied should never be empty its a prefilled database field. Please make sure your database has been updated properly or fix it by reinstalling the last update from my repo. Does the config report you got less then expected fields?

**Comment by jayamin-10** [comment]
Hey sorry for late reply. i get this message at the top of the configuration page : "Phpsaml expected 21 configuration items but got 19 items instead"
i didn't get this part 'Please make sure your database has been updated properly or fix it by reinstalling the last update from my repo' are you talking about the latest version of the plugin from your rep? I did that but it's not working with guest id. The same error everytime"PHP Warning: Undefined array key "proxied".

**Comment by jayamin-10** [comment]
Hey @DonutsNL hope you're doing well.
![proxied](https://github.com/derricksmith/phpsaml/assets/62603670/1192f666-62d0-474d-b32c-6432bc5ae84e)

This is proxied? Look, this is how it displays on my screen. It doesn't allow me to put/select anything. Also i get this message at the top of the configuration page "Phpsaml expected 21 configuration items but got 19 items instead"
Thanks in advance for your reply.

**Comment by jayamin-10** [comment]
@DonutsNL Thanks a lot for your help. the manual approach solved the proxied problem. now i can select 'yes' or 'no' options.
But couldn't yet login with a guest account. The same blank screen.
Here are the errors i got from glpi error log : 
PHP Warning:  Undefined array key "http://schemas.xmlsoap.org/xx/xxxx/xx/identity/claims/surname" in /glpi/plugins/phpsaml/inc/phpsaml.class.php on line 348, referer: https://login.microsoftonline.com/
PHP Warning:  Trying to access array offset on value of type null in /glpi/plugins/phpsaml/inc/phpsaml.class.php on line 349, referer: https://login.microsoftonline.com/
PHP Warning:  Undefined array key "http://schemas.xmlsoap.org/xx/xxxx/xx/identity/claims/givenname" in /glpi/plugins/phpsaml/inc/phpsaml.class.php on line 350, referer: https://login.microsoftonline.com/
PHP Warning:  Trying to access array offset on value of type null in /glpi/plugins/phpsaml/inc/phpsaml.class.php on line 350, referer: https://login.microsoftonline.com/
PHP Warning:  Undefined variable $error in /glpi/plugins/phpsaml/inc/phpsaml.class.php on line 320, referer: https://login.microsoftonline.com/

**Comment by DonutsNL** [comment]
The required claims are missing in the SAML response from the idp. Are these claims added in the response?

**Comment by jayamin-10** [comment]
@DonutsNL Yes they were provided. These surname and givenname errors are gone apart from this error:
PHP Warning:  Undefined variable $error in /glpi/plugins/phpsaml/inc/phpsaml.class.php on line 320, referer: https://login.microsoftonline.com/
PHP Warning:  Undefined variable $error in /glpi/plugins/phpsaml/inc/phpsaml.class.php on line 320, referer: https://login.microsoftonline.com/

**Comment by DonutsNL** [comment]
Hi @jayamin-10 

I still expect that not all required claims are provided and is causing the JIT user->add to fail. 

I added:

- Additional checks to the JIT function to make sure you end up in an errorstate if stuff is missing;
- Additional logging that should enable us to review the errorlog and See what claims are provided and missing;
- Redirects to make sure you dont end up on a blank page;
- Additional UI messages hopefully explaining whats going on without the need of reviewing the logfiles.

Please download and install the latest version from my repo and tell me what the messages/log say now.

**Comment by jayamin-10** [comment]
It redirects back to the login page and does not register any error on glpi error log file for a guest user. But like before it works with a user of the domain.
The cmails provided are :
![image](https://github.com/derricksmith/phpsaml/assets/62603670/0271fade-c620-4271-9f3b-ab5842e9f4f5)

**Comment by DonutsNL** [comment]
let me quickly write a procedure to dump the provided Saml post to a local file so we can review that.

**Comment by DonutsNL** [comment]
Changed the front/acs.php it should now dump the headers in the /front/ directory for you to review. 

Please logon using both users and review the dumps. Then please tell me what is different. Dont share that file anywhere for obvious reasons :-)

**Comment by jayamin-10** [comment]
@DonutsNL Hello again, 
I did the tests. The dump files are identical for the normal account on the domain and guest account.
but some of the attribute values  were different such as: 
1. The attribute : https://schemas.microsoft.com/identity/claims/identityprovider
=>  I got  : https://sts.windows.net/xxxxxxx-xxxxxx-xxxxx-xxxx-xxxxxxxxxx for **normal account** 
    and for guest account i just got the word "mail'
2. For http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name i got the value as my.email@address.com for normal account but for the guest account i got : saml_plugin_yahoo.com#EXT#@mydomain.onmicrosoft.com
The same is true for nameid attribute.

**Comment by DonutsNL** [comment]
Hmm, i need to dive into the code. I am not sure i would want to allow login using these fields. 

Out of curiocity is the correct account information (email, idp) available somewhere in the returned SAMLResponse?

**Comment by DonutsNL** [comment]
In addition, it seems Microsoft will allow you to transform claim information of guests when fields are empty. See this document: https://learn.microsoft.com/en-us/azure/active-directory/develop/saml-claims-customization

**Comment by jayamin-10** [comment]
Yes the correct email address for normal account is present on the nameid, emailaddress attributes but for guest account saml_plugin@yahoo.com is only present for emailaddress. 
here is the attribute link where i get correct email address for both account : http://schemas.xmlsoap.org/xx/xx/identity/claims/emailaddress
by idp you mean the idp i mentioned in the other comment or i should have got another idp value ?

i didn't mention in the first comment : i get other attributes aswell like surname, givenname, displayname etc.

**Comment by DonutsNL** [comment]
If you trust me enough, could you send a example of the SAMLReponse to: chris [dot] gralike [at] flevo-scouts [dot] nl? 

This so i can review its structure and see if I can add 'allow guests' to the plugin.

**Comment by jayamin-10** [comment]
@DonutsNL after transforming the nameid attribute claim with user.mail (test.saml@yahoo.com) instead of user.principalname (test.saml_yahoo.com#EXT#@domain.onmicrosoft.com) we can now sign in with the guest account. 
I think glpi doesn't accept ".com#EXT#" as ID
Thanks a lot. you have been a great help. 
For me the plugin works.

**Comment by DonutsNL** [comment]
Glad it is working.
Ill add an additional check in the acs and write a warning pointing out to the MS documentation

**Comment by DonutsNL** [comment]
I added an additional check to detect a guest user without transformations, could validate its working for me?

**Comment by jayamin-10** [comment]
Hey @DonutsNL i will let you know tomorrow about the detection of a guest user without trasformation of nameid claim. Btw it would be really kind of you, if you could fix  the bug for proxied where we had to pass three sql query manually to alter table. A lot of people would benefit from it. on that note i haven't quite get the point of proxied because i tested with proxied yes and no but didn't notice any difference.

**Comment by DonutsNL** [comment]
Hey @jayamin-10 ,

Yeah its something I need to dive into. Its also the reason why I added the fieldcheck in the config page. This to allow someone to correct it if the DB is corrupted for whatever reason. On the proxy my understanding is that it breaks the URLs used for redirect and validation in some cases because the plugin is looking at the wrong HTTP headers. There is a param for this in the PHP saml plugin i need to config.

**Comment by jayamin-10** [comment]
Hey again @DonutsNL,

I did the test. this is the warning message we get. 
![image](https://github.com/derricksmith/phpsaml/assets/62603670/ecbdb334-23c6-4d33-9a9e-42fde14e2a57)

I think this is good,. People will have an idea how to solve the issue. If you could just add in the warning "Solution : use the value user.mail  for the the claims 'nameid' and 'name' instead of using the default value user.principalname on your Azure sso application"

here is how i transformed the the claims which works : 
![image](https://github.com/derricksmith/phpsaml/assets/62603670/f1b976cb-0dfc-4a50-a04f-9f503c3307cb)

**Comment by DonutsNL** [comment]
Hi @jayamin-10,

Thank you for validating my check. Good it is working as intended. 
I updated the error message with your suggestion.

**Comment by DonutsNL** [comment]
@derricksmith, i think some merges will prove fruitfull. Also this issue can be closed.
